### PR TITLE
Lite: Batch_to_space-ND support

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
Arbitrary dimension support added.

Add ref: #21526

### CODE REVIEW

**Comment by jdduke** [comment]
See also https://github.com/tensorflow/tensorflow/pull/27867. Once we confirm these 2 PRs unblock the model execution, we can proceed.

**Comment by gbaned** [comment]
@ANSHUMAN87 Can you please resolve conflicts? Thanks!

**Comment by ANSHUMAN87** [comment]
@gbaned : The conflict is resolved, would you please help get it reviewed and merged, Thanks!

**Comment by miaout17** [comment]
> See also #27867. Once we confirm these 2 PRs unblock the model execution, we can proceed.

Thanks for the contribution, @ANSHUMAN87 
Any reply to jdduke's previous comment?

**Comment by ANSHUMAN87** [comment]
> > See also #27867. Once we confirm these 2 PRs unblock the model execution, we can proceed.
> 
> Thanks for the contribution, @ANSHUMAN87
> Any reply to jdduke's previous comment?

@miaout17 : I have verified, the solution works fine! Please refer #27867 also, both are done together as related. Thanks!

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by gbaned** [comment]
@ANSHUMAN87 Can you please resolve conflicts? Thanks!

**Comment by ANSHUMAN87** [comment]
Resolved!

**Comment by gbaned** [comment]
@miaout17 Can you please take a look on this PR? Thanks!

**Comment by gbaned** [comment]
@ANSHUMAN87 Can you please resolve conflicts? Thanks!

**Comment by ANSHUMAN87** [comment]
@gbaned : The conflicts are resolved now, Thanks!

**Comment by gbaned** [comment]
@ANSHUMAN87 Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by ANSHUMAN87** [comment]
@gbaned : I will resolve the conflicts possibly by Saturday. Thanks!

**Comment by ANSHUMAN87** [comment]
@gbaned : The conflicts are resolved now. Thanks!

File: tensorflow/lite/kernels/batch_to_space_nd.cc [hunk 1]
```diff
@@ -98,12 +94,65 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
...
-                 NumDimensions(op_context.input) >= kInputMinDimensionNum);
-  TF_LITE_ENSURE(context,
-                 NumDimensions(op_context.input) <= kInputMaxDimensionNum);
+
+  // Input dimension should be minimum [batch] + spatial_shape
+  TF_LITE_ENSURE(context, NumDimensions(op_context.input) >
+                              NumElements(op_context.block_shape) + 1);
   TF_LITE_ENSURE_EQ(context, op_context.input->type, op_context.output->type);
 
+  // Currently only int32 is supported for block_shape &  crops
```
Comment (thaink) [inline tensorflow/lite/kernels/batch_to_space_nd.cc:103]: Please add dot at the end of comments.

File: tensorflow/lite/kernels/batch_to_space_nd.cc [hunk 2]
```diff
@@ -48,47 +48,43 @@ struct BatchToSpaceNDContext {
...
-                                 crops[dim * 2] - crops[dim * 2 + 1];
+
+  int prod_block_shape = 1;
+  for (int dim = 0; dim < SizeOfDimension(op_context->block_shape, 0); ++dim) {
+    TF_LITE_ENSURE(context, crops[dim * 2] >= 0);
+    TF_LITE_ENSURE(context, crops[dim * 2 + 1] >= 0);
+    TF_LITE_ENSURE(context, block_shape[dim] >= 1);
+    int final_dim_size = (input_size->data[dim + 1] * block_shape[dim] -
+                          crops[dim * 2] - crops[dim * 2 + 1]);
+    // Should be non-negative
```
Comment (thaink) [inline tensorflow/lite/kernels/batch_to_space_nd.cc:76]: I think this comment is excessive.

File: tensorflow/lite/kernels/batch_to_space_nd.cc [hunk 3]
```diff
@@ -48,47 +48,43 @@ struct BatchToSpaceNDContext {
...
   }
 
-  output_size->data[0] = output_batch_size;
-  output_size->data[input_size->size - 1] =
-      input_size->data[input_size->size - 1];
+  // Number of batch must be multiple of
+  // (block_shape[0] * ...  * block_shape[M]).
+  TF_LITE_ENSURE_EQ(context, input_size->data[0] % prod_block_shape, 0);
+
+  // output_batch_size =  [batch / prod(block_shape)]
```
Comment (thaink) [inline tensorflow/lite/kernels/batch_to_space_nd.cc:86]: This comment can be removed too.

**Comment by thaink** [comment]
As a high-level comment, is it possible to define local variables of type NdArrayDesc<8> instead of introducing new temporary tensors? We are unlikely to support dimensions > 8.

**Comment by thaink** [comment]
I found a performance regression with your PR on my Pixel 3.
The input size is [400,30,30,100 ] and batch_size is [20, 20], crops  is [[1,1],[1,1]].
The performance before and after your pr:
before 57.826ms
after 99.597ms
Could you double check on performance?

**Comment by jdduke** [comment]
Hey @ANSHUMAN87, we're working on an overhaul for >4D kernel support, so I think it's best if we table this PR for now (and 1 or 2 of the related PRs) as we work toward that end. Thanks for your patience in iterating on this and pushing for generalization, and apologies we couldn't land this more promptly.

**Pull request closed without merging**